[Spark-15155][Mesos] Optionally ignore default role resources

hellertime opened the pull request:
Add new boolean property: `spark.mesos.ignoreDefaultRoleResources`. When this property is set Spark will only accept resources from the role passed in the `spark.mesos.role` property. If `spark.mesos.role` has not been set, `spark.mesos.ignoreDefaultRoleResources` has no effect.

Additional unit tests added to `MesosSchedulerBackendSuite`, extending the original multi-role test suite.

hellertime committed:
[Spark-15155][Mesos] Optionally ignore default role resources

Add new boolean property: `spark.mesos.ignoreDefaultRoleResources`.
When this property is set Spark will only accept resources from the role passed in the `spark.mesos.role` property.
If `spark.mesos.role` has not been set, `spark.mesos.ignoreDefaultRoleResources` has no effect.

Additional unit tests added to `MesosSchedulerBackendSuite`, extending the original multi-role test suite.

hellertime commented:
@tnachen I'd like to request you review this change, since you are the author of the original mesos role code.

AmplabJenkins commented:
Can one of the admins verify this patch?

tnachen commented:
So from what I can see, what you really want is to only don't select resources from * role right? 
Otherwise if you're getting offers from other roles you are already registered for those roles as well, which begs the question why are you registered with that role as well.

tnachen commented:
jenkins please test

tnachen commented:
ok to test

SparkQA commented:
**[Test build #57891 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57891/consoleFull)** for PR 12933 at commit [`48b4723`](https://github.com/apache/spark/commit/48b472329036e62ef00973f912d54db7fc5d7872).

SparkQA commented:
**[Test build #57891 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57891/consoleFull)** for PR 12933 at commit [`48b4723`](https://github.com/apache/spark/commit/48b472329036e62ef00973f912d54db7fc5d7872).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57891/
Test FAILed.

hellertime commented:
@tnachen exactly, I want to keep Spark from grabbing * roles, as in my use case I have a particular spark cluster that I want to isolate from other clusters.

SparkQA commented:
**[Test build #57905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57905/consoleFull)** for PR 12933 at commit [`8419664`](https://github.com/apache/spark/commit/84196648410ffdf9d51900559abfe0ef83764580).

SparkQA commented:
**[Test build #57905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57905/consoleFull)** for PR 12933 at commit [`8419664`](https://github.com/apache/spark/commit/84196648410ffdf9d51900559abfe0ef83764580).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57905/
Test FAILed.

hellertime commented:
*MiMa* test died here:
```
/home/jenkins/workspace/SparkPullRequestBuilder@2/dev/mima: line 37: 18083 Aborted                 (core dumped) java -XX:MaxPermSize=1g -Xmx2g -cp "$TOOLS_CLASSPATH:$OLD_DEPS_CLASSPATH" org.apache.spark.tools.GenerateMIMAIgnore
[error] running /home/jenkins/workspace/SparkPullRequestBuilder@2/dev/mima -Pyarn -Phadoop-2.3 -Pkinesis-asl -Phive-thriftserver -Phive ; received return code 134
Attempting to post to Github...
 > Post successful.
```

I don't think this is due to my change, but I'm really not familiar with the *MiMa* test.

tnachen commented:
I see, I wonder besides the wildcard role if it ever makes sense to have two lists of roles, one set that you are registered with and another set you want to accept resources for, I cannot think of any possible scenario.
Perhaps we should just introduce one boolean flag that default takes wildcard role resources or not? spark.mesos.rejectWildcardRole or something better named?

hellertime commented:
I see that as a potential solution, however at some future point mesos will support frameworks which can hold multiple roles (MESOS-1763 😉) so perhaps leaving it in this form will ease that transition (though we'd then need to change the other property to `spark.mesos.roles`, so really its a wash).

tnachen commented:
Yes I am aware of multi roles, but not sure in what situation you would want to register with multiple roles but only use one role's resources?

hellertime commented:
True. So really the desired behavior is always just to ignore `*` resources. A boolean property would suffice here. What about `spark.mesos.ignoreDefaultResourceRole`, and have the property ignored if `spark.mesos.role` is unset?

hellertime commented:
I'm updating the title of the PR to reflect the change in approach. I think the boolean property will be sufficient.

SparkQA commented:
**[Test build #57989 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57989/consoleFull)** for PR 12933 at commit [`5ea2f46`](https://github.com/apache/spark/commit/5ea2f46909db99bed0a78fe3f2dc7af316087c36).

hellertime commented:
Alright, I've modified the code to convert from a property which enumerated the accepted roles, to one which will simply ignore the default role when it is set.

I also removed one unit test, which was redundant with the test that checks we accepts all roles when `spark.mesos.role` is set.

SparkQA commented:
**[Test build #57989 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57989/consoleFull)** for PR 12933 at commit [`5ea2f46`](https://github.com/apache/spark/commit/5ea2f46909db99bed0a78fe3f2dc7af316087c36).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57989/
Test FAILed.

hellertime commented:
Rebasing against master.

SparkQA commented:
**[Test build #57993 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57993/consoleFull)** for PR 12933 at commit [`d2b7ad4`](https://github.com/apache/spark/commit/d2b7ad444e02b947f4a7264018b4e48610731408).

SparkQA commented:
**[Test build #57993 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57993/consoleFull)** for PR 12933 at commit [`d2b7ad4`](https://github.com/apache/spark/commit/d2b7ad444e02b947f4a7264018b4e48610731408).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57993/
Test FAILed.

tnachen commented:
@hellertime Can you rebase and submit again to rerun the tests?

tnachen commented:
@hellertime ping

hellertime commented:
Hi! Was busy with the day job. I didn't mean to let this slip! Absolutely will rebase and retest. Thanks.

Sent from my iPhone

> On Jun 21, 2016, at 11:16 PM, Timothy Chen <notifications@github.com> wrote:
> 
> @hellertime ping
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

tnachen commented:
@hellertime Are you able to rebase?

hellertime commented:
@tnachen rebased with master

SparkQA commented:
**[Test build #66740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66740/consoleFull)** for PR 12933 at commit [`e81ff73`](https://github.com/apache/spark/commit/e81ff73a53eeb7eed645f3042ae9da758a254179).

SparkQA commented:
**[Test build #66740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66740/consoleFull)** for PR 12933 at commit [`e81ff73`](https://github.com/apache/spark/commit/e81ff73a53eeb7eed645f3042ae9da758a254179).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66740/
Test FAILed.

hellertime commented:
Fixed scala style error

SparkQA commented:
**[Test build #66741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66741/consoleFull)** for PR 12933 at commit [`181f207`](https://github.com/apache/spark/commit/181f2075199549571364224f45914831b49f3310).

SparkQA commented:
**[Test build #66741 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66741/consoleFull)** for PR 12933 at commit [`181f207`](https://github.com/apache/spark/commit/181f2075199549571364224f45914831b49f3310).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66741/
Test FAILed.

SparkQA commented:
**[Test build #66780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66780/consoleFull)** for PR 12933 at commit [`838dc77`](https://github.com/apache/spark/commit/838dc77d5473e7a584efbd3ac223eba696a427f7).

SparkQA commented:
**[Test build #66780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66780/consoleFull)** for PR 12933 at commit [`838dc77`](https://github.com/apache/spark/commit/838dc77d5473e7a584efbd3ac223eba696a427f7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66780/
Test FAILed.

SparkQA commented:
**[Test build #66811 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66811/consoleFull)** for PR 12933 at commit [`bbe908e`](https://github.com/apache/spark/commit/bbe908ef66898624166bcf06d1773008e7414f14).

SparkQA commented:
**[Test build #66811 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66811/consoleFull)** for PR 12933 at commit [`bbe908e`](https://github.com/apache/spark/commit/bbe908ef66898624166bcf06d1773008e7414f14).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66811/
Test FAILed.

SparkQA commented:
**[Test build #66825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66825/consoleFull)** for PR 12933 at commit [`81e07c8`](https://github.com/apache/spark/commit/81e07c876faab4cb9b9327df5ca9749c52e3cdc2).

SparkQA commented:
**[Test build #66825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66825/consoleFull)** for PR 12933 at commit [`81e07c8`](https://github.com/apache/spark/commit/81e07c876faab4cb9b9327df5ca9749c52e3cdc2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66825/
Test FAILed.

hellertime commented:
@tnachen I'm not sure what to do about this unit test failure. Running `./dev/run-tests` on my system does not produce the error, and trying to just run the mesos suite with `./build/mvn -DwildcardSuites=org.apache.spark.scheduler.cluster.mesos test` results in a successful build, it doesn't appear to actually run the test. Any advice on how to test this locally?

tnachen commented:
I just tried running it locally and I'm getting the same error. It seems like with your change that test is simply declining the offer.

hellertime commented:
You saw the error with `./dev/run-tests`? Ok I'll figure this out.

Sent from my iPhone

> On Oct 13, 2016, at 12:24 AM, Timothy Chen <notifications@github.com> wrote:
> 
> I just tried running it locally and I'm getting the same error. It seems like with your change that test is simply declining the offer.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

SparkQA commented:
**[Test build #66887 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66887/consoleFull)** for PR 12933 at commit [`d64b261`](https://github.com/apache/spark/commit/d64b26171c013772c680333c8c742e4965df3b1d).

SparkQA commented:
**[Test build #66887 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66887/consoleFull)** for PR 12933 at commit [`d64b261`](https://github.com/apache/spark/commit/d64b26171c013772c680333c8c742e4965df3b1d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66887/
Test PASSed.

tnachen commented:
I ran `mvn test` inside of the mesos folder.

On Thu, Oct 13, 2016 at 3:21 AM, Chris Heller <notifications@github.com>
wrote:

> You saw the error with `./dev/run-tests`? Ok I'll figure this out.
>
> Sent from my iPhone
>
> > On Oct 13, 2016, at 12:24 AM, Timothy Chen <notifications@github.com>
> wrote:
> >
> > I just tried running it locally and I'm getting the same error. It seems
> like with your change that test is simply declining the offer.
>
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/12933#issuecomment-253474465>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEGrAU4TKCiLBwMpDm7PmGsEm79tdoEks5qzgY0gaJpZM4IYBAd>
> .
>

tnachen commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -50,6 +50,44 @@ trait MesosSchedulerUtils extends Logging {
+      case _ => Set(Some("*"), role)
+    }
+    val acceptedRoles = roles.flatten
+    logDebug(s"Accepting resources from role(s): ${acceptedRoles.mkString(",")}")
```
I think we should move this log outside of this helper method, as there might be other context in the future that's calling this.

tnachen commented:
@srowen can you help review this? Besides my minor comment overall it looks fine with me.

srowen commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala:
```diff
@@ -551,12 +551,16 @@ private[spark] class MesosClusterScheduler(
-        getResource(o.resources, "cpus") >= driverCpu &&
-        getResource(o.resources, "mem") >= driverMem
+        val acceptableResources = o.resources.asScala
+          .filter((r: Resource) => acceptedResourceRoles(r.getRole))
```
Just a style thing, but I think you'd generally just write `(r => acceptedResourceRoles(r.getRole))`

srowen commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala:
```diff
@@ -551,12 +551,16 @@ private[spark] class MesosClusterScheduler(
+          .filter((r: Resource) => acceptedResourceRoles(r.getRole))
+          .asJava
+        getResource(acceptableResources, "cpus") >= driverCpu &&
+        getResource(acceptableResources, "mem") >= driverMem
```
Indent this continuation 2 spaces to make it clear it's a continuation

srowen commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -50,6 +50,44 @@ trait MesosSchedulerUtils extends Logging {
+   */
+  protected def getAcceptedResourceRoles(props: Map[String, String]): Set[String] = {
+    getAcceptedResourceRoles(
+      props.get("spark.mesos.ignoreDefaultRoleResources") match {
```
This might be my taste only but I'm finding this style hard to read. 
`props.get(...).map(_.toBoolean).getOrElse(false)`? It's what I'd expect to see in the spark code base but don't know if it's objectively any better. Still I'd break this out into a `val` rather than squeeze this into a method arg

srowen commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:
```diff
@@ -50,6 +50,44 @@ trait MesosSchedulerUtils extends Logging {
+  private def getAcceptedResourceRoles(
+      ignoreDefaultRoleResources: Boolean,
+      role: Option[String]) = {
+    val roles = ignoreDefaultRoleResources match {
```
This seems more readable:

```
val roles = if (ignoreDefaultRoleResources && role.isDefined) Set(role) else Set(Some("*"), role)
```

srowen reviewed (changes requested):
I can only usefully comment on style and left a few comments

HyukjinKwon commented:
Hi @hellertime, this PR seems inactive for few months after the last review comments above. Would this be better closed if you are not currently able to work on this further maybe?
